### ISSUE TITLE
Impossible to switch scenes, lots of crash situations. Editor workflow.

### ISSUE THREAD

**Issue by Alan-FGR** [AtomicGameEngine/AtomicGameEngine#1245]
After investigation on why @toddgibson wasn't being able to switch scenes, I tried to achieve that and failed miserable, ended up discovering some problems:

Ref: https://files.gitter.im/AtomicGameEngine/AtomicGameEngine/tLTP/blob

As reported by @toddgibson, LoadScene() doesn't discard the current scene, but I'm not sure if that's the intended behavior. Since he wanted to effectively 'switch' scenes, I started looking for a way to manually remove the previous scene in a component.

Cadaver said somewhere that Scene.LoadXML should remove all previous data, but that also doesn't happen. I've also tried storing the previous scene and calling Clear(), Remove(), GetNode().Remove(), SetEnabled(false) and all of these simply crash the engine. Either when called before or after loading the next scene, cached or not.

I thought that was because the Viewport wasn't switching scenes properly but even just a GetSubsystem<Viewport>().SetScene(yourScene) also crashes the engine.

Even if you just disable the camera on your first scene and call LoadScene() to load another scene, the engine will also crash.

Confirmed for loading scenes from C# component. Need confirmation for other languages.

Reproduction steps: Create a new project in the editor, in the default scene disable the camera, create another scene and try to load it from a component in the first scene, compile and play. Optionally, re-enable the camera in the first scene, if you play now everything works, but the first scene can't be removed no matter what.

**Comment by Alan-FGR** [comment]
I've tried adding a bool to make sure the code runs only once so it's not a problem with Update() running and altering the scene while it's still being loaded or something.

**Comment by JoshEngebretson** [comment]
Both the C# and C++ feature examples create and teardown scenes from a top level example menu:

https://github.com/AtomicGameEngine/AtomicExamples/tree/master/FeatureExamples

There is also player api which will load scenes.  It would be good to setup a scene loading/switching example which defines best practices for handling this and catch/refine issues related to it.

**Comment by toddgibson** [comment]
Yes, I agree a scene loading/switching example would be helpful.  As a new user, looking at the FeatureExamples project wasn't exactly clear because it only includes one ".scene" file.  So having a separate example with multiple .scene files should prove useful.

**Comment by Alan-FGR** [comment]
@JoshEngebretson I don't think this is only a matter of creating an example... switching to a new scene shouldn't crash if the previous one don't have a camera enabled for example, this is consistently reproducible using the editor. In my opinion, even if you do something you're not supposed to, a crash should never happen without at least an error. While it's true that creating scenes programmatically works fine, something is definitely wrong in the editor workflow, I don't know what is and don't have much time to check, maybe something wrong in the .scene files the editor spits out, or maybe it's just some calls you have to make before unloading the old and loading a new scene, like setting up the camera and viewport. In any case, I still have no idea on how to create two scenes using the editor and switch between them using components without the engine crashing.

**Comment by shaddockh** [comment]
In a project I've been working on,  I was able to switch scenes programmatically by loading up a scene file and then setting it on the viewport.  

What I've noticed is that ```Atomic.player.loadScene``` will load up and display a scene only the first time.  After that, you need to manipulate the current viewport.

This is a very basic example and will attach a new scene to the provided viewport and then activate that scene's camera.  It's simple, but could help with a starting point for something much more solid.  

Basically what I've done is the following:
```typescript
   let currentScene: Atomic.Scene = null;
   function switchToScene(sceneName: string, viewPortIndex = 0) {
        const firstScene = this.currentScene === null;
        const scene = Atomic.player.loadScene(sceneName);
        if (!firstScene) {
            const viewPort = Atomic.renderer.getViewport(viewPortIndex);
            viewPort.setScene(scene);
            viewPort.setCamera(scene.getMainCamera());
        }
        currentScene = scene;
    }
```

**Comment by Alan-FGR** [comment]
@shaddockh Thanks, I think I got it. I can try to make the example this weekend.

**Comment by Alan-FGR** [comment]
Guys, any ideas on how to reliably get the main camera on other languages?
Also, can we only trust the reference dropping on C# side to 'collect' the previous scene? I think yes because when it goes out of scope it's collected even when it's active but I'm not 100% sure if you shouldn't manually call something, even if only as a good practice.

**Comment by Alan-FGR** [comment]
Any ideas? Don't you have to call Remove() on the previous scene manually? If so, I couldn't find a way to do that properly, but I'm not 100% sure if that's necessary, when the GC collects the previous scene I guess the destructor should take care of that for you, but what I'm struggling a bit to understand is the relation to the viewport and the camera in the scene. If you just switch the viewport it keeps using the camera from the previous scene, but shouldn't it be collected? At this point there's no reference to if from the managed side so I don't know why the camera doesn't get destroyed.

I'm basically doing what @shaddockh pointed out: calling LoadScene(), then on the viewport calling SetScene() and SetCamera(). The goal here is to get rid of the previous scene data altogether.

**Comment by JoshEngebretson** [comment]
The View has a reference to the current Camera.  

The Player class isn't very fleshed out.  It can pretty much load a scene, give you a current scene, and that is about it: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicPlayer/Player.h#L45

I am currently working on some scene (re)loading issues from this test: https://github.com/LumaDigital/AtomicTests/tree/MB_CSharpMemoryLeaksTest/CSharpTests/MemoryLeaks

I would until https://github.com/AtomicGameEngine/AtomicGameEngine/tree/JME-ATOMIC-OBJECTMETRICS lands, which will fix the scene load issue and may punch up the Player subsystem.

Otherwise, the C# feature example switches between scenes of the multidemo by simply building up the scenes and setting the viewport.

**Comment by Alan-FGR** [comment]
OK, wait... the camera is attached to a node in a scene... but the viewport has a reference to it, so when you call LoadScene the scene gets destroyed except for the camera node? That's the part I don't quite understand. If the camera is an object in the scene, how it comes it stays there when you load a new scene?
Anyway, is it safe to just call LoadScene to load a new scene and change the scene and camera in the current viewport? Is that guaranteed to destroy the previous scene without manual intervention? (as long as it goes outta scope and you're not keeping references somewhere obviously)

**Comment by JoshEngebretson** [comment]
```master``` now has some new Player subsystem API and fixes some C# scene lifetime issues: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicPlayer/Player.h#L45

There is a test case which uses the new API here: https://github.com/AtomicGameEngine/AtomicTests/tree/master/CSharpTests/MemoryLeaks

The new object tracking should help catch objects leaking: https://github.com/AtomicGameEngine/AtomicGameEngine/wiki/Profiling

**Comment by Alan-FGR** [comment]
OK, I'm kinda drowning in stuff to do, but I'm going to take a look this weekend. I should've sent what I've done so far...

**Comment by Alan-FGR** [comment]
Well, how am I supposed to pass the camera from the new scene in LoadScene? That only makes sense if you're managing cameras separately I guess, because the idea here is to switch to another scene and just forget about the previous stuff and use the camera from the new scene returned in LoadScene... That being said, doesn't LoadScene destroy all previous objects? If so, how are you supposed to keep the camera to pass there? I don't think using that is applicable for this example anyway, but I'm curious on what would be the use case. I could see for example how a bool to switch to a camera in the new scene automatically could be useful though.

**Comment by Alan-FGR** [comment]
OK, there's still some leak going on, each scene switch the memory usage increases considerably (>10MB).
I'm just doing this:
```
Scene oldScene = GetSubsystem<Player>().CurrentScene;
Scene newScene = GetSubsystem<Player>().LoadScene($"Scenes/{button.Text}.scene");
                
Vector<Camera> cameras = new Vector<Camera>();
newScene.GetComponents(cameras, true);

Viewport viewport = GetSubsystem<Renderer>().GetViewport(0);
viewport.SetScene(newScene);
viewport.SetCamera(cameras[0]);

//oldScene.Remove(); gets nulld
```
I can profile it later to try and detect exactly what is going on.

**Comment by JoshEngebretson** [comment]
Scene derives from Node, calling ```scene.Remove();``` is a nop unless somehow you have it parented to another Node ;)

The basic Player subsystem supports having multiple scenes loaded at a time:

https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicPlayer/Player.h#L45

You can unload a scene with ```UnloadScene```: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicPlayer/Player.h#L61

Once you do this, there shouldn't be any further manipulation of outstanding scene references.  Let me know if the API is lacking something...

**Comment by Alan-FGR** [comment]
OK, I guess UnloadScene is what I need there. Also, there is a bit of misinformation here I should clarify: in fact, it's only when you use LoadXML() to load contents in a Scene that the previous stuff gets destroyed. If you're creating multiple scenes of course they won't affect each other. That's the part I wasn't understanding. So, since LoadScene() creates a new Scene object, that's surely the expected outcome and it won't make sense if the scenes affected each other. It's still not clear when scenes should be GC'd though, because if you don't keep a reference on the managed side they get collected, but in this case they don't.

**Comment by Alan-FGR** [comment]
Hi. Sorry the long delay. I still couldn't sort this out...
https://gitter.im/AtomicGameEngine/AtomicGameEngine?at=5889328811e7a7f61de63916
Creating scenes manually from code works fine. Loading .scene resources in components is proving to be way more challenging than one would expect though.

**Comment by Alan-FGR** [comment]
Unloading the Scene solved the leak (it has to be unloaded before loading the next one otherwise it gets rid of the viewport). How should I deliver that example? It seems a bit overkill to have a whole example in the front page grid while it could be just a script example somewhere...

**Comment by shaddockh** [comment]
Not everything in the Atomic Examples repo shows up on the grid.  Maybe you could store it under https://github.com/AtomicGameEngine/AtomicExamples/tree/master/Cookbook or make a wiki entry.

**Comment by Alan-FGR** [comment]
I guess we'll have to go with the entire project for this example.

Since examples should be easy to execute I was considering making a component in which you'd drag-n-drop your scenes in the inspector and it would display a menu with proper switching functionality, but apparently there's no way to have a field where you can drop scenes or even generic objects from the project and get the file path in the component code.

In my opinion a cookbook example could be appropriated *only if* it's OK to create the scenes in code (which wouldn't specifically address the request in this issue). It would also be necessary to know what's the current scene, and ideally with more than debug text. Are there built-in primitive meshes like cubes/cylinders we could use? If not, maybe using different clear colors could help, although it's not an elegant solution. The only drawback is we won't catch problems caused by the XML scene loader for example, but I honestly don't think that's a big deal.

In the mean time I'm porting what I have here to JS and TS.

**Comment by Alan-FGR** [comment]
Anybody wants to port the C# code to TS/JS? It's the only thing left... @shaddockh? :P

**Comment by shaddockh** [comment]
Sure, I can take a look.  Where is the C# example located?

**Comment by Alan-FGR** [comment]
https://gist.github.com/Alan-FGR/bd0b60c5f9aad5edac301508e748b2b9
Since it's a cookbook example I'm not using a component. I've tested here and it's leak-free.
@shaddockh I'll have to comment the code sooner or later so please let me know if you want to wait for the comments before porting... totally forgot about that :P

**Comment by Alan-FGR** [comment]
Also, naming is a bit odd, specifically sampleTypes

**Comment by shaddockh** [comment]
No need for comments.  That example looks pretty straightforward and should be easy to port to JS/TS.
